runtime/errors/errors.go

```diff
@@ -66,3 +106,107 @@ type ParentError interface {
 	error
 	ChildErrors() []error
 }
+
+// MemoryError indicates a memory limit has reached and should end
+// the Cadence parsing, checking, or interpretation.
+type MemoryError struct {
+	Err error
+}
+
+func (e MemoryError) Unwrap() error {
+	return e.Err
+}
+
+func (e MemoryError) Error() string {
+	return fmt.Sprintf("memory error: %s", e.Err.Error())
+}
+
+func (e MemoryError) IsUserError() {}
+
+// UnexpectedError is the default implementation of InternalError interface.
+// It's a generic error that wraps an implementation error.
+//
+type UnexpectedError struct {
```

Comment:
What is the difference between an unexpected error and an unreachable error? Is it possible for an unexpected error to occur with, e.g. a malformed FVM, while unreachable errors can never occur at all?

Reply:
`UnexpectedError` is a generic error to be thrown in any unhandled cases. It's essentially a wrapper for go-errors, to make sure all errors that are thrown are known 'cadence errors'.
i.e: replaces all existing `fmt.Errorf()` / `errors.new()`.

`UnreachableError` doesn't wrap anything. it only prints the stack trace.

Reply:
oh sorry, I didn't see the last part of the question.
Both of these errors should never occur. Errors caused by malformed FVM are wrapped as `ExternalError` (this was already there).

Reply:
I was also wondering what the difference between the two types is. Maybe we can merge them, or alternatively, I guess an `UnreachableError` could be wrapped by/nested in an `UnexptectedError`? ("it was unexpected this code path was reached" – compared with e.g. another `UnexpectedError` which wraps a concrete "unreachable error", like in `importValue`)

Reply:
The `UnexpectedError` was introduced to make go-errors conform to the `InternalError` interface. Given that `UnreachableError` already implements the `InternalError`, wrapping it with `UnexpectedError` doesn't add much value – Only adds complexity to the error chain.

So I merged the two (https://github.com/onflow/cadence/pull/1718/commits/098d47b5ebf124b25d120248c509fa5fa40624b6) as I suggested earlier. So the `UnexpectedError` has both a cause (error) and the stack trace – and the trace info is available for all scenarios.

Now then, this makes the `UnreachableError` redundant, so removed it, and simply used a built-in error string.

Reply:
Nice!